### PR TITLE
Forbid block morph in CSE.

### PR DESCRIPTION
**Pull request by sandreenko**
https://github.com/dotnet/runtime/pull/33845/commits/77f235c2c9c5ebef4ff10b491377cc6fd64bf292 forbids calling `fgMorphCopyBlock` or `fgMorphInitBlock` if we are in CSE phase. These functions could delete trees that are CSE candidates, which leads to asserts later. Exposed via struct changes, without retyping we have more structs and more CSE uses/defs.
There are no size diffs from that change, only a few text diffs, that look like changes in instruction order,
before:
```
IN0047: 0001AB lea      rsi, bword ptr [V02+0x20 rsp+A8H]
IN0048: 0001B3 mov      rbp, gword ptr [V02 rsp+88H]
IN0049: 0001BB lea      rdi, bword ptr [rbp+232]
```
after:
```
IN0047: 0001AB mov      rbp, gword ptr [V02 rsp+88H]
IN0048: 0001B3 lea      rdi, bword ptr [rbp+232]
IN0049: 0001BA lea      rsi, bword ptr [V02+0x20 rsp+A8H]
```
it happens because calls to `fgMorphBlockOperand` wraps everything in a `IND(ADDR(X))`, but if you call `fgMorphTree` on such a tree again it will fold them back to `X`.  So without the change we were adding useless `IND(ADDR())` nodes, now we don't (in CSE phase, others still do that, I tried to fix that behavior but it took me too far).
before:
```
N309 (  3,  4) [000212] ------------       t212 =    LCL_FLD_ADDR byref  V02 loc0         u:2[+32] Fseq[_currentToken] rsi REG rsi
                                                  /--*  t212   byref  
N311 (  6,  7) [000366] nc----------       t366 = *  IND       struct REG NA
```
after:
```
N311 (  3,  4) [000210] ------------       t210 =    LCL_FLD   ref    V02 loc0         u:2[+0] Fseq[_scanner] rbp REG rbp <l:$21c, c:$21d>
```

The second commit changes how we dump struct types in `; Initial local variable assignments` table:
before
```
;  V00 arg0              ref  class-hnd
;  V01 arg1           struct (16) 
;  V02 loc0             long 
;  V03 loc1            byref  pinned
;  V04 OutArgs        lclBlk (na)  "OutgoingArgSpace"
```
after
```
;  V00 arg0              ref  class-hnd
;  V01 arg1           struct <System.Span`1[Char], 16>
;  V02 loc0             long 
;  V03 loc1            byref  pinned
;  V04 OutArgs        lclBlk <na>  "OutgoingArgSpace"
```

I was planning to change it in the final local variable assignments printing too, but first: it generates too many asm diffs, second - right now it is nicely formatted as a table and adding long names there destroys this formatting.

### CODE REVIEW

**Comment by sandreenko** [comment]
PTAL @CarolEidt @dotnet/jit-contrib

**Review by CarolEidt — APPROVED** [review]
LGTM - nice to see such a straightforward fix! (Also like the dump changes)

File: src/coreclr/src/jit/morph.cpp [hunk 1]
```diff
@@ -13774,14 +13781,6 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
                 break;
```
Comment (briansull) [inline src/coreclr/src/jit/morph.cpp]: This is badly formatted.

I believe it should be written:

```
    if (tree->OperIsCopyBlkOp())
    {
        break;
    }
```

File: src/coreclr/src/jit/morph.cpp [hunk 2]
```diff
@@ -13757,6 +13757,13 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
     switch (oper)
     {
         case GT_ASG:
+            // Make sure we're allowed to do this.
+            if (optValnumCSE_phase)
+            {
+                // It is not safe to reorder/delete CSE's
+                break;
+            }
+
```
Comment (briansull) [inline src/coreclr/src/jit/morph.cpp]: How does reordering this change the behavior here?

It seems to me that now we will break here, but before we would have also always done a break

Comment (sandreenko) [reply src/coreclr/src/jit/morph.cpp]: GitHub doesn't show you the real change, 
image

File: src/coreclr/src/jit/morph.cpp (hunk 1 above)
Comment (sandreenko) [reply src/coreclr/src/jit/morph.cpp]: yeah, if you have a change after the 10000th line than GitHub shows you all incorrect. We have #13359 but it is not clear how to separate these files without doing morp1.cpp, morph2.cpp.

**Pull request merged**